src/nvim/main.c

```diff
@@ -1801,117 +1796,121 @@ static void exe_commands(mparm_T *parmp)
   TIME_MSG("executing command arguments");
 }
 
-/*
- * Source startup scripts.
- */
-static void source_startup_scripts(mparm_T *parmp)
+/// Source vimrc or do other user initialization
+///
+/// Does one of the following things, stops after whichever succeeds:
+///
+/// 1. Execution of VIMINIT environment variable.
+/// 2. Sourcing user vimrc file ($XDG_CONFIG_HOME/nvim/init.vim).
+/// 3. Sourcing other vimrc files ($XDG_CONFIG_DIRS[1]/nvim/init.vim, …).
+/// 4. Execution of EXINIT environment variable.
+///
+/// @return True if it is needed to attempt to source exrc file according to
+///         'exrc' option definition.
+static bool do_user_initialization(void)
+  FUNC_ATTR_WARN_UNUSED_RESULT
 {
-  int i;
+  bool do_exrc = p_exrc;
+  if (process_env("VIMINIT", true) == OK) {
+    return do_exrc;
+  }
+  char_u *user_vimrc = (char_u *)stdpaths_user_conf_subpath("init.vim");
+  if (do_source(user_vimrc, true, DOSO_VIMRC) != FAIL) {
+    if (do_exrc) {
+      do_exrc = (path_full_compare((char_u *)VIMRC_FILE, user_vimrc, false)
+                 != kEqualFiles);
+    }
+    xfree(user_vimrc);
+    return do_exrc;
+  }
+  xfree(user_vimrc);
+  char *const config_dirs = stdpaths_get_xdg_var(kXDGConfigDirs);
+  if (config_dirs != NULL) {
+    const void *iter = NULL;
+    do {
+      const char *dir;
+      size_t dir_len;
+      iter = vim_colon_env_iter(config_dirs, iter, &dir, &dir_len);
+      if (dir == NULL || dir_len == 0) {
+        break;
+      }
+      const char path_tail[] = { 'n', 'v', 'i', 'm', PATHSEP,
+                                 'i', 'n', 'i', 't', '.', 'v', 'i', 'm', NUL };
+      char *vimrc = xmalloc(dir_len + sizeof(path_tail) + 1);
+      memmove(vimrc, dir, dir_len);
+      vimrc[dir_len] = PATHSEP;
+      memmove(vimrc + dir_len + 1, path_tail, sizeof(path_tail));
```

Comment:
though it doesn't matter here, ARRAY_SIZE would be a "best practice" instead of sizeof.